library/bigip_ssl_certificate.py

```diff
@@ -0,0 +1,449 @@
+                return False
+            else:
+                raise F5ModuleError(e)
+        else:
+            return True
+
+    def exists_cert(self, name):
+        return self.api.tm.sys.crypto.certs.cert.exists(name=name)
+
+    def exists_key(self, name):
+        return self.api.tm.sys.crypto.keys.key.exists(name=name)
+
+    def import_cert(self, **kwargs):
+        name = kwargs['name']
+        cert_format = kwargs['cert_format']
+        if cert_format == 'pem':
+            try:
+                certfilename = kwargs['certfilename']
+            except KeyError:
+                certfilename = kwargs['name'] + '.crt'
+            else:
+                self.upload(certfilename)
+                self.install_cert(name, os.path.basename(certfilename))
+            try:
+                keyfilename = kwargs['keyfilename']
+            except KeyError:
+                pass
+            else:
+                self.upload(keyfilename)
+                self.install_key(name, os.path.basename(keyfilename))
+            return True
+        elif cert_format == 'pkcs12':
+            raise F5ModuleError("pkcs12 not supported by f5-sdk")
+
+    def install_cert(self, name, certfilename):
+        filename = os.path.join('/var/config/rest/downloads', certfilename)
+        params = {'from-local-file': filename, 'name': name}
+        self.api.tm.sys.crypto.certs.exec_cmd('install', **params)
+
+    def install_key(self, name, keyfilename):
+        filename = os.path.join('/var/config/rest/downloads', keyfilename)
+        params = {'from-local-file': filename, 'name': name}
+        self.api.tm.sys.crypto.keys.exec_cmd('install', **params)
+
+    def upload(self, filename):
+        self.api.shared.file_transfer.uploads.upload_file(filename)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(type='str', default='present',
+                       choices=['absent', 'present']),
+            user=dict(type='str', required=True, aliases=['username']),
+            password=dict(type='str', required=True, no_log=True),
+            server=dict(type='str', required=True),
+            name=dict(type='str', required=True),
+            partition=dict(type='str', default='Common'),
+            connection=dict(type='str', default='rest',
+                            choices=['icontrol', 'rest'],
+                            aliases=['transport']),
+            cert_format=dict(type='str', default='pem',
+                             choices=['pem']),
+            cert_pem_file=dict(type='str'),
+            key_pem_file=dict(type='str'),
+            pkcs12_file=dict(type='str'),
+            pkcs12_password=dict(type='str', no_log=True),
+            validate_cert=dict(type='bool', default=True),
+            # overwrite=dict(type='bool', default=False)
+        )
+    )
+    result = {'changed': False}
+    if module.params['connection'] == 'icontrol':
```

Comment:
I've read the file upload does not reliably work until LTM v12.
Specifically, the /var/config/rest/download/tmp directory keeps deleting
itself(at an interval I haven't figured out yet) and needs to be manually
created.  I left the SOAP calls in place for people still on v11.  If that
is not desirable, it can be removed.

On Mon, Jun 13, 2016 at 12:23 PM, Tim Rupp <notifications@github.com> wrote:

> In library/bigip_ssl_certificate.py
> <https://github.com/F5Networks/f5-ansible/pull/18#discussion_r66822785>:
>
> > +            partition=dict(type='str', default='Common'),
> > +            connection=dict(type='str', default='rest',
> > +                            choices=['icontrol', 'rest'],
> > +                            aliases=['transport']),
> > +            cert_format=dict(type='str', default='pem',
> > +                             choices=['pem']),
> > +            cert_pem_file=dict(type='str'),
> > +            key_pem_file=dict(type='str'),
> > +            pkcs12_file=dict(type='str'),
> > +            pkcs12_password=dict(type='str', no_log=True),
> > +            validate_cert=dict(type='bool', default=True),
> > +            # overwrite=dict(type='bool', default=False)
> > +        )
> > +    )
> > +    result = {'changed': False}
> > +    if module.params['connection'] == 'icontrol':
>
> If we can do without SOAP, then you can eliminate the entire statement
> here and just use the REST call
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/F5Networks/f5-ansible/pull/18/files/b156b3c9d2e8dd85246b216eb7c28aeda45f7a6b#r66822785>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALqwB6NI5frdamwZ8SsQ0Ws-Cd5USGr0ks5qLYQIgaJpZM4I0cLm>
> .
>